_blogposts/2023-05-25-dynamic-import.mdx

```diff
@@ -0,0 +1,138 @@
+
+[Dynamic import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import), a feature allowing for asynchronous code loading in JavaScript, has long offered potential for code splitting, lazy loading, and via that reduced initial load times for applications. While ReScript has always supported dynamic imports, they've often been cumbersome to maintain in their current form. We're excited to announce that, with ReScript v11, we're introducing first-class support for dynamic imports - making them ergonomic to use and easy to maintain.
+
+But before we dive into how dynamic imports look in v11, let's do a short primer on what the challenges has been when using dynamic imports in ReScript prior to v11.
+
+## The current state of dynamic imports in ReScript
+
+Dynamic imports in JavaScript work on _paths_. This means that if you want to dynamically import the file `src/utils/text/markdown.js` from `src/components/text/MarkdownRenderer.js`, you'd need to spell the full relative path out:
+
+```js
+// MarkdownRenderer.js
+let markdown = await import("../../../utils/text/markdown.js");
+```
+
+In JavaScript, modules and files are referred to by their file system path, which makes sense considering its design. However, in ReScript, all files represent modules that are globally available for anyone to use, without having to care about where that file is located on the file system.
+
+This global availability of modules is generally beneficial, as it simplifies file restructuring and relieves developers from the need to remember the precise location of each file. However, when it comes to dynamic imports, this poses a challenge.
+
+As a ReScript developer, you're not accustomed to tracking the relative file system locations of your modules. But with dynamic imports, you'd would have to manually keep track of where your files are located relative to each other, in order for your dynamic import binding to point to the right file.
+
+Moreover, in ReScript you can change what suffix your generated files has. It's `.bs.js` by default, but it could be `.mjs`, `.bs.mjs` or even just `.js` depending on what project and environment you're in. This flexibility is another great feature of ReScript, but it further complicates dynamic imports as the import paths need to account for the file suffixes.
+
+Before v11, these factors made maintaining dynamic imports in ReScript quite challenging. Let's take a look at how we've addressed these issues in v11.
+
+## Importing a value
+
+Dynamically importing a value, like a function, is now done via the new `Js.import` function. You pass a reference to the value you want to dynamically import, just as if you'd use the value directly. `Js.import` will return a promise resolving to the value you pass into it, that you can then `await`. The compiler will ensure that the JavaScript file holding that value is not referenced directly, but rather is dynamically imported.
```

Comment:
Just wanted to point out one potential ambiguity in the text. The phrase "The compiler will ensure that the JavaScript file holding that value is not referenced directly, but rather is dynamically imported" could be read differently by some. "Direct reference" might be interpreted by some readers as a static import, like let add2 = MathUtils.add, instead of specifying the file path in a dynamic import statement.

Perhaps it could be clearer by saying something like: "The compiler will ensure that the JavaScript file holding that value is not dynamically imported via a string literal containing the file path, but instead by directly referencing the module or value."

Reply:
I dropped most of this text in favour of showing the example. When comparing the input / output, it should be clear what's happening.